videostore/utils.go

```diff
@@ -229,41 +263,85 @@ func formatDateTimeToString(dateTime time.Time) string {
 
 // matchStorageToRange returns a list of files that fall within the provided time range.
 // Includes trimming video files to the time range if they overlap.
-func matchStorageToRange(files []string, start, end time.Time, logger logging.Logger) []string {
+// The matcher assumes that the input files list is sorted by start time and the underylying
+// video segments do not overlap.
+func matchStorageToRange(files []fileWithDate, start, end time.Time, logger logging.Logger) []string {
 	var matchedFiles []string
-	for _, file := range files {
-		dateTime, err := extractDateTimeFromFilename(file)
-		if err != nil {
-			logger.Debugf("failed to extract datetime from filename: %s, error: %v", file, err)
-			continue
+	// Cache of the first matched video file's width, height, and codec
+	// to ensure every video in the matched files set have the same params.
+	var firstSeenVideoInfo videoInfo
+	// Find the first file to consider for matching. First search for the first file that starts after the query start time.
+	// We then want to consider the previous file as well, since it may overlap with the query start time.
+	var firstFileIndex int
+	firstFileIndexFound := false
+	for i, file := range files {
+		if file.date.After(start) {
+			firstFileIndex = i - 1
+			firstFileIndexFound = true
+			break
+		}
+	}
+	// If no file starts after the query start time, we still want to consider the last file.
+	if !firstFileIndexFound {
+		firstFileIndex = len(files) - 1
+	}
+	if firstFileIndex < 0 {
+		firstFileIndex = 0
+	}
+	// Iterate through the files and find the ones that match the time range.
+	for i := firstFileIndex; i < len(files); i++ {
+		fileStartTime := files[i].date
+		fileName := files[i].name
+		// If the file starts after the query end time, we can stop searching.n
+		if fileStartTime.After(end) {
+			logger.Debugf("Skipping file %s and winding down matcher. File starts after end time (start=%v, end=%v)", fileName, start, end)
+			break
 		}
-		duration, err := getVideoDuration(file)
+		videoFileInfo, err := getVideoInfo(fileName)
 		if err != nil {
-			logger.Debugf("failed to get video duration for file: %s, error: %v", file, err)
+			logger.Debugf("failed to get video duration for file: %s, error: %v", fileName, err)
 			continue
 		}
-		fileEndTime := dateTime.Add(duration)
-		// Check if the file's time range intersects with [start, end)
-		if dateTime.Before(end) && fileEndTime.After(start) {
+		actualFileEndTime := fileStartTime.Add(videoFileInfo.duration)
+		// Check if the segment file's time range intersects
+		// with the match request time range [start, end)
+		if fileStartTime.Before(end) && actualFileEndTime.After(start) {
```

Comment:
do we need this if statement?

Didn't we already establish that every element we are iterating over is after the start time and before the end time?